File: src/CPP/src/pratice/LetterCombinationsOfAPhoneNumber.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

// [Source]: http://www.1point3acres.com/bbs/thread-225716-1-1.html
// http://www.1point3acres.com/bbs/thread-222632-1-1.html
// https://leetcode.com/problems/letter-combinations-of-a-phone-number/

class SolutionDFS {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty())
            return {};

        unordered_map<char, string> digitMap;
        digitMap['0'] = " ";
        digitMap['1'] = "";
        digitMap['2'] = "abc";
        digitMap['3'] = "def";
        digitMap['4'] = "ghi";
        digitMap['5'] = "jkl";
        digitMap['6'] = "mno";
        digitMap['7'] = "pqrs";
        digitMap['8'] = "tuv";
        digitMap['9'] = "wxyz";

        vector<string> res;
        vector<char> partial;
        helper(digits, 0, digitMap, partial, res);
        return res;
    }

    void helper(string &digits, size_t idx, unordered_map<char, string> &digitMap, vector<char> &partial, vector<string> &res) {
        if (idx == digits.length()) {
            res.push_back(string(partial.begin(), partial.end()));
            return;
        }

        string &candidates = digitMap[digits[idx]];
        if (candidates.empty()) { // num 1
            helper(digits, idx + 1, digitMap, partial, res);
        } else {
            for (char &c : candidates) {
                partial.push_back(c);
                helper(digits, idx + 1, digitMap, partial, res);
                partial.pop_back();
            }
        }
    }
};

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty())
            return {};

        unordered_map<char, string> digitMap;
        digitMap['0'] = " ";
        digitMap['1'] = "";
        digitMap['2'] = "abc";
        digitMap['3'] = "def";
        digitMap['4'] = "ghi";
        digitMap['5'] = "jkl";
        digitMap['6'] = "mno";
        digitMap['7'] = "pqrs";
        digitMap['8'] = "tuv";
        digitMap['9'] = "wxyz";

        string partial = "";
        queue<string> prevRes;
        prevRes.push(partial);
        for (int idx = 0; idx < (int)digits.length(); ++idx) {
            string &candidates = digitMap[digits[idx]];
            if (candidates.empty())
                continue;

            int size = prevRes.size();
            for (int i = 0; i < size; ++i) {
                string &prev = prevRes.front();
                prevRes.pop();

                for (char &c : candidates) {
                    prevRes.push(prev + c);
                }
            }
        }

        vector<string> res;
        while (!prevRes.empty()) {
            res.push_back(prevRes.front());
            prevRes.pop();
        }
        return res;
    }
};

int main() {
    Solution sol;

    for (string &str : sol.letterCombinations("123"))
        cout << str << endl;

    return 0;
}

File: wwwroot/js/letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    let res = [];
    if (digits === null || digits.length === 0) {
        return res;
    }

    appendChar(digits, 0, [], res);
    return res;
};

const digitMap = {
    '2': "abc",
    '3': "def",
    '4': "ghi",
    '5': "jkl",
    '6': "mno",
    '7': "pqrs",
    '8': "tuv",
    '9': "wxyz"
};

/**
 * @param {string} digits
 * @param {int} idx
 * @param {string[]} currChars
 * @param {string[]} res
 */
const appendChar = function(digits, idx, currChars, res) {
    if (idx === digits.length) {
        res.push(currChars.join(''));
        return;
    }

    let poss = digitMap[digits[idx]];
    for (let i = 0; i < poss.length; i++) {
        currChars.push(poss[i]);
        appendChar(digits, idx + 1, currChars, res);
        currChars.pop();
    }
};

document.getElementById("letter-combinations-of-a-phone-number").innerHTML = letterCombinations("23");